### PR TITLE
feature/ add event category remove endpoint

### PR DESCRIPTION
**Pull request by Draktolli**
https://trello.com/c/0Ty19kl0/200-eventservice-add-eventcategory-remove-endpoint

### CODE REVIEW

File: src/EventService.Data/CategoryRepository.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+using System.Threading.Tasks;
+using LT.DigitalOffice.EventService.Data.Interfaces;
+using LT.DigitalOffice.EventService.Data.Provider;
+using Microsoft.EntityFrameworkCore;
+
+namespace LT.DigitalOffice.EventService.Data;
+
+public class CategoryRepository : ICategoryRepository
+{
+  private readonly IDataProvider _provider;
```
Comment (liza459) [inline src/EventService.Data/CategoryRepository.cs]: You forgot the constructor.

File: src/EventService.Data/Interfaces/ICategoryRepository.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using LT.DigitalOffice.Kernel.Attributes;
+
+namespace LT.DigitalOffice.EventService.Data.Interfaces;
+
+[AutoInject]
+public interface ICategoryRepository
+{
+  public Task<bool> DoesExistAsync(List<Guid> categoryIds);
```
Comment (liza459) [inline src/EventService.Data/Interfaces/ICategoryRepository.cs]: Remove "public" in method interface.
Check all  interface.

File: src/EventService.Business/Commands/EventCategory/RemoveEventCategoryCommand.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    if (!validationResult.IsValid)
+    {
+      return _responseCreator.CreateFailureResponse<bool>(
+        HttpStatusCode.BadRequest,
+        validationResult.Errors.Select(er => er.ErrorMessage).ToList());
+    }
+
+    await _repository.RemoveAsync(request.EventId, request.EventCategoryIds);
+
```
Comment (liza459) [inline src/EventService.Business/Commands/EventCategory/RemoveEventCategoryCommand.cs:66]: You need to assign a status code

**Review by liza459 — COMMENTED** [review]
You need to add an endpoint to the controller

**Comment by vorobevamv** [comment]
> You need to add an endpoint to the controller

+1

File: src/EventService.Validation/EventCategory/RemoveEventCategoryRequestValidator.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+{
+  public RemoveEventCategoryRequestValidator(
+    IEventRepository eventRepository,
+    ICategoryRepository categoryRepository,
+    IEventCategoryRepository eventCategoryRepository)
+  {
+    RuleFor(request => request.EventId)
+      .MustAsync(async (x, _) => await eventRepository.DoesExistAsync(x))
+      .WithMessage("This event doesn't exist");
+    RuleFor(request => request.EventCategoryIds)
```
Comment (vorobevamv) [inline src/EventService.Validation/EventCategory/RemoveEventCategoryRequestValidator.cs]: add empty line

File: src/EventService.Validation/EventCategory/RemoveEventCategoryRequestValidator.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+      .MustAsync(async (x, _) => await categoryRepository.DoesExistAsync(x))
+      .WithMessage("Some categories doesn't exist.");
+    RuleFor(request => request)
```
Comment (vorobevamv) [inline src/EventService.Validation/EventCategory/RemoveEventCategoryRequestValidator.cs:25]: add empty line

File: src/EventService.Validation/EventCategory/RemoveEventCategoryRequestValidator.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+      .MustAsync(async (x, _) => await eventCategoryRepository.DoesExistAsync(x.EventId, x.EventCategoryIds))
+      .WithMessage("No such categories in this event.");
```
Comment (vorobevamv) [inline src/EventService.Validation/EventCategory/RemoveEventCategoryRequestValidator.cs]: "This event doesn't belong to all categories in the list."

File: src/EventService.Data/EventCategoryRepository.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  }
+
+  public async Task<bool> DoesExistAsync(Guid eventId, List<Guid> categoryIds)
+  {
+    return await _provider.EventsCategories.AsNoTracking().AnyAsync(ec => categoryIds.Contains(ec.CategoryId) && ec.EventId == eventId);
+  }
+
+  public async Task<bool> RemoveAsync(Guid eventId, List<Guid> categoryIds)
+  {
+    if (categoryIds is null)
```
Comment (vorobevamv) [inline src/EventService.Data/EventCategoryRepository.cs]: `if (categoryIds is null || !categoryIds.Any())`

File: src/EventService.Data/CategoryRepository.cs (hunk 1 above)
Comment (TIove) [reply src/EventService.Data/CategoryRepository.cs]: Lemme guess, you didn't test it? J

File: src/EventService.Data/EventCategoryRepository.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    {
+      return false;
+    }
+
+    IQueryable<DbEventCategory> eventCategories = _provider.EventsCategories.Where(ec => categoryIds.Contains(ec.CategoryId) && ec.EventId == eventId);
```
Comment (TIove) [inline src/EventService.Data/EventCategoryRepository.cs]: Check this through profiler, is this good query?

File: src/EventService.Validation/EventCategory/RemoveEventCategoryRequestValidator.cs [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+public class RemoveEventCategoryRequestValidator : AbstractValidator<RemoveEventCategoryRequest>, IRemoveEventCategoryRequestValidator
+{
+  public RemoveEventCategoryRequestValidator(
+    IEventRepository eventRepository,
+    ICategoryRepository categoryRepository,
+    IEventCategoryRepository eventCategoryRepository)
+  {
+    RuleFor(request => request.EventId)
+      .MustAsync(async (x, _) => await eventRepository.DoesExistAsync(x))
+      .WithMessage("This event doesn't exist");
```
Comment (TIove) [inline src/EventService.Validation/EventCategory/RemoveEventCategoryRequestValidator.cs]: Stop in the end of text

File: src/EventService.Business/Commands/EventCategory/RemoveEventCategoryCommand.cs [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    return new OperationResultResponse<bool> { Body = true };
```
Comment (vorobevamv) [inline src/EventService.Business/Commands/EventCategory/RemoveEventCategoryCommand.cs]: what if  `_repository.RemoveAsync(request.EventId, request.EventCategoryIds)` returns 'false'?

File: src/EventService.Data/CategoryRepository.cs [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+  public async Task<bool> DoesExistAsync(List<Guid> categoryIds)
```
Comment (TIove) [inline src/EventService.Data/CategoryRepository.cs]: Hmm, name is not clear. Reading this name I cannot get that can pass List of categories. My suggestions:
* Use params
* or Rename method to DoesExistAllAsync e.g., or smth like this

File: src/EventService.Data/EventCategoryRepository.cs [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    _provider.EventsCategories.RemoveRange(eventCategories);
```
Comment (vorobevamv) [inline src/EventService.Data/EventCategoryRepository.cs]: may be remove variable 'eventCategories'?

File: src/EventService.Validation/EventCategory/RemoveEventCategoryRequestValidator.cs [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    RuleFor(request => request.EventCategoryIds)
+      .MustAsync(async (x, _) => await categoryRepository.DoesExistAsync(x))
+      .WithMessage("Some categories doesn't exist.");
```
Comment (TIove) [inline src/EventService.Validation/EventCategory/RemoveEventCategoryRequestValidator.cs:23]: What if request.EventCategoryIds is null?

File: src/EventService.Business/Commands/EventCategory/RemoveEventCategoryCommand.cs [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+        validationResult.Errors.Select(er => er.ErrorMessage).ToList());
+    }
+
+    OperationResultResponse<bool> response = new();
+    response.Body = await _repository.RemoveAsync(request.EventId, request.EventCategoryIds);
+
+    if (!response.Body)
+    {
+      return _responseCreator.CreateFailureResponse<bool>(HttpStatusCode.BadRequest);
+    }
```
Comment (liza459) [inline src/EventService.Business/Commands/EventCategory/RemoveEventCategoryCommand.cs:63]: add empty line

**Review by liza459 — APPROVED** [review]

File: src/EventService/Controllers/EventCategoryController.cs [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+using Microsoft.AspNetCore.Mvc;
+
+namespace LT.DigitalOffice.EventService.Controllers
+{
+  [Route("[controller]")]
+  [ApiController]
+  public class EventCategoryController : ControllerBase
+  {
+    [HttpDelete("remove")]
+    public async Task<OperationResultResponse<bool>> DeleteAsync(
```
Comment (vorobevamv) [inline src/EventService/Controllers/EventCategoryController.cs]: RemoveAsync

File: src/EventService.Data/EventCategoryRepository.cs [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+  private readonly IDataProvider _provider;
+
+  public EventCategoryRepository(IDataProvider provider)
+  {
+    _provider = provider;
+  }
+
+  public bool DoesExistAsync(Guid eventId, List<Guid> categoryIds)
+  {
+    return categoryIds.All(categoryId => _provider.EventsCategories.AnyAsync(ec => ec.CategoryId == categoryId && ec.EventId == eventId).Result);
```
Comment (vorobevamv) [inline src/EventService.Data/EventCategoryRepository.cs]: move '_provider.EventsCategories...' to next line (with tab)

File: src/EventService.Data/CategoryRepository.cs [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+  private readonly IDataProvider _provider;
+
+  public CategoryRepository(IDataProvider provider)
+  {
+    _provider = provider;
+  }
+
+  public bool DoesExistAllAsync(List<Guid> categoryIds)
+  {
+    return categoryIds.All(categoryId => _provider.Categories.AsNoTracking().AnyAsync(c => c.Id == categoryId && c.IsActive).Result);
```
Comment (vorobevamv) [inline src/EventService.Data/CategoryRepository.cs]: and here. move '_provider.Categories...' to next line

File: src/EventService.Data/EventCategoryRepository.cs [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  public async Task<bool> RemoveAsync(Guid eventId, List<Guid> categoryIds)
+  {
+    if (categoryIds is null || !categoryIds.Any())
+    {
+      return false;
+    }
+
+    _provider.EventsCategories.RemoveRange(_provider.EventsCategories.Where(ec => categoryIds.Contains(ec.CategoryId) && ec.EventId == eventId));
```
Comment (vorobevamv) [inline src/EventService.Data/EventCategoryRepository.cs]: move '_provider.EventsCategories.Where...' to next line

File: src/EventService.Data/Interfaces/IEventCategoryRepository.cs [hunk 19]
```diff
@@ -0,0 +1,13 @@
...
+using System.Threading.Tasks;
+using LT.DigitalOffice.Kernel.Attributes;
+
+namespace LT.DigitalOffice.EventService.Data.Interfaces;
+
+[AutoInject]
+public interface IEventCategoryRepository
+{
+  bool DoesExistAsync(Guid eventId, List<Guid> categoryIds);
+  Task<bool> RemoveAsync(Guid eventId, List<Guid> categoryIds);
```
Comment (vorobevamv) [inline src/EventService.Data/Interfaces/IEventCategoryRepository.cs]: add empty line (between methods)

**Review by vorobevamv — APPROVED** [review]

**Review by Segway-svg — APPROVED** [review]

**Review by TheDioBrando — APPROVED** [review]

File: src/EventService.Data/EventCategoryRepository.cs [hunk 20]
```diff
@@ -31,13 +31,28 @@ public async Task<bool> CreateAsync(List<DbEventCategory> dbEventCategories)
...
-    return categoryIds.All(categoryId => _provider.EventsCategories.AnyAsync(ec => ec.CategoryId == categoryId && ec.EventId == eventId).Result);
+    return categoryIds.All(categoryId =>
+      _provider.EventsCategories.AnyAsync(ec => ec.CategoryId == categoryId && ec.EventId == eventId).Result);
   }
 
-  public Task<int> CountCategoriesAsync(Guid eventId)
+  public async Task<bool> RemoveAsync(Guid eventId, List<Guid> categoryIds)
   {
-    return _provider.EventsCategories.AsNoTracking().CountAsync(ec => ec.EventId == eventId);
+    if (categoryIds is null || !categoryIds.Any())
```
Comment (VLiamin) [inline src/EventService.Data/EventCategoryRepository.cs]: May be CategoriesIds?

File: src/EventService.Business/Commands/EventCategory/RemoveEventCategoryCommand.cs [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+    ValidationResult validationResult = await _validator.ValidateAsync(request);
+
+    if (!validationResult.IsValid)
+    {
+      return _responseCreator.CreateFailureResponse<bool>(
+        HttpStatusCode.BadRequest,
+        validationResult.Errors.Select(er => er.ErrorMessage).ToList());
+    }
+
+    OperationResultResponse<bool> response = new();
```
Comment (VLiamin) [inline src/EventService.Business/Commands/EventCategory/RemoveEventCategoryCommand.cs]: You can write this like this
image

**Review by VLiamin — APPROVED** [review]

**Pull request merged**